### PR TITLE
Enhanced metadata cache management

### PR DESCRIPTION
**Pull request by cpeel**
@c-w I'm opening this pull request up mostly to generate discussion about what I've been working on. It needs more testcases as well as more thorough testing of the sqlalchemy+sqlite backend and corner cases.

Background: I'm working to incorporate the Gutenberg package into OpenLibrary to highlight PG ePubs for OpenLibrary records that have PG numbers associated with them. An inability to cleanly specify the cache location and having the cache populate on first use don't really fit into such a service-oriented. This branch is working on encapsulating the metadata cache logic into its own class, giving the package user much more control over how it works.

As noted in the commit, this is backwards compatible with the exception that the cache must be manually populated the first time, otherwise it throws an exception. This is to ensure the element of least surprise ('cause an 18-hour return time on the first call is very surprising)!

### CODE REVIEW

**Comment by c-w** [comment]
See also #31.

**Comment by c-w** [comment]
Thanks for the pull request. Looks like a great start.

I'm not at all wedded to keeping compatibility with BSD-DB (makes the installation more of a hassle, issues on Windows, etc), so feel free to entirely remove BSD-DB in favour of SQLite.

**Comment by cpeel** [comment]
Ok, this is as good as I want to make it right now. Unfortunately after getting all the work in to enable use with SQLAlchemy+sqlite it turns out the BSD backend is 3 times faster than the SQLAlchemy+sqlite backend during loading.

Even worse, cache population using SQLAlchemy+sqlite segfaults on a trimmed-down PG catalog with 2024 entries in it (BSD backend completes in 153 seconds on the same dataset). It's certainly possible that pointing SQLAlchemy towards a real DB might be faster and succeed on larger data sets, but that seems very heavy-handed for this use-case.

I think the metadata cache manager is still immensely useful for more fine-grained control over the backend cache, particularly when being used by a web service, but the use of SQLAlchemy+sqlite is not a silver bullet for making it faster. I suspect for that we'd have to move off of an RDF backend altogether and onto a backend (possibly sqlite) with a custom schema.

**Comment by cpeel** [comment]
@c-w, Is this work still of interest to you? Getting a more fine-grained level of control over the creation and management of the cache is blocking some of PG integration work into openlibrary, so I'm hoping to get this, or something like it, in.

If there are additional things you'd like to see in this branch, please let me know.

**Comment by c-w** [comment]
@cpeel: Thanks for the pull request and apologies for the delay in getting back to you. Given that all the tests pass and that you added new test, I'm happy to merge this branch. I'll take a deeper look after I once again have more time to devote to this project.

File: gutenberg/acquire/metadata.py [hunk 1]
```diff
@@ -26,23 +28,148 @@
...
-    data_url = r'http://www.gutenberg.org/cache/epub/feeds/rdf-files.tar.bz2'
-    with tempfile.NamedTemporaryFile(delete=False) as metadata_archive:
-        shutil.copyfileobj(urllib2.urlopen(data_url), metadata_archive)
-    yield metadata_archive.name
-    remove(metadata_archive.name)
+_GUTENBERG_CATALOG_URL = \
+    r'http://www.gutenberg.org/cache/epub/feeds/rdf-files.tar.bz2'
+
+
+class CacheAlreadyExistsException(Exception):
```
Comment (c-w) [inline gutenberg/acquire/metadata.py]: @cpeel: What do you think about replacing these custom exceptions with `ValueError`? The exceptions can only be raised on one context so that using a built-in exception type is unambiguous.

Comment (cpeel) [inline gutenberg/acquire/metadata.py]: If we replaced them with ValueErrors we'd need to check the value of the exception string to be able to tell them apart from other possible ValueError exceptions raised to calls like get_metadata() -- that seems unpythonic. Introducing a root exception class to derive them from makes sense though.

File: gutenberg/acquire/metadata.py [hunk 2]
```diff
@@ -26,23 +28,148 @@
...
+    def __init__(self, store, cache_uri):
+        self.identifier='urn:gutenberg:metadata'
+        if store == 'Sleepycat':
+            self.store = store
+            self.removable = True
+        else:
+            if cache_uri.startswith('sqlite://'):
+                self.removable = True
+            else:
+                self.removable = False
```
Comment (c-w) [inline gutenberg/acquire/metadata.py]: @cpeel: Is support for non-sqlite/non-removable graphs something on which you depend? There are no tests covering this code branch so I'd assume that this is not a feature that you are using and can therefore safely be removed. Could you please confirm this?

File: gutenberg/acquire/metadata.py (hunk 1 above)
Comment (c-w) [inline gutenberg/acquire/metadata.py]: Thanks for your input. As of [9f3aff2](https://github.com/c-w/Gutenberg/blob/9f3aff2/gutenberg/_domain_model/exceptions.py) we now have `gutenberg.Error` as the top-level class for all exceptions thrown by the API.

File: gutenberg/acquire/metadata.py (hunk 2 above)
Comment (cpeel) [inline gutenberg/acquire/metadata.py]: This is not a feature on which I depend.

While untested, the current code should work with other SQLAlchemy backends (like mysql, pgsql, etc) by using the appropriate cache_uri. I had originally anticipated testing the performance of different SQLAlchemy backends beyond sqlite, but currently believe that any SQLAlchemy backend will be less performant than the BSDDB.
